src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// primary workspace for remote host. no one except solution service can update this workspace
+    /// </summary>
+    internal class RemoteWorkspace : Workspace
+    {
+        public const string WorkspaceKind_RemoteWorkspace = "RemoteWorkspace";
+
+        public RemoteWorkspace()
+            : base(RoslynServices.HostServices, workspaceKind: RemoteWorkspace.WorkspaceKind_RemoteWorkspace)
+        {
+            PrimaryWorkspace.Register(this);
+
+            Options = Options.WithChangedOption(CacheOptions.RecoverableTreeLengthThreshold, 0);
+        }
+
+        public override bool CanApplyChange(ApplyChangesKind feature)
+        {
+            // apply change is not allowed
+            return false;
+        }
+
+        public override bool CanOpenDocuments
+        {
+            get
+            {
+                // enables simulation of having documents open.
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Clears all projects and documents from the workspace.
+        /// </summary>
+        public new void ClearSolution()
+        {
+            base.ClearSolution();
+        }
+
+        /// <summary>
+        /// Adds an entire solution to the workspace, replacing any existing solution.
+        /// </summary>
+        public Solution AddSolution(SolutionInfo solutionInfo)
+        {
+            if (solutionInfo == null)
```

Comment:
I made it to use lock just in case. since like adhocworkspace, this workspace is controlled by us, there is only one who can update the workspace. but just to be safe. added lock.